File: demos_code_J2_conception_objet/07_package_stability_metrics/DemoStability.java
import java.util.List;

public class DemoStability {
    record PackageStats(String name, int ca, int ce) {
        double instability() {
            return (ca + ce) == 0 ? 0.0 : (double) ce / (ca + ce);
        }
    }

    public static void main(String[] args) {
        List<PackageStats> packages = List.of(
                new PackageStats("domain", 8, 1),
                new PackageStats("application", 4, 4),
                new PackageStats("stripe-adapter", 0, 6),
                new PackageStats("shared", 11, 9)
        );

        System.out.printf("%-16s %4s %4s %8s  %s%n", "package", "Ca", "Ce", "I", "lecture");
        for (PackageStats p : packages) {
            double i = p.instability();
            String reading = i < .3 ? "stable" : i > .7 ? "instable" : "intermédiaire";
            System.out.printf("%-16s %4d %4d %8.2f  %s%n", p.name(), p.ca(), p.ce(), i, reading);
        }
    }
}

File: Demo en Python/demos_code_J2_conception_objet_python/07_package_stability_metrics/demo_stability.py
from dataclasses import dataclass


@dataclass(frozen=True)
class PackageStats:
    name: str
    ca: int
    ce: int

    def instability(self) -> float:
        total = self.ca + self.ce
        return 0.0 if total == 0 else self.ce / total


if __name__ == "__main__":
    packages = [
        PackageStats("domain", 8, 1),
        PackageStats("application", 4, 4),
        PackageStats("stripe-adapter", 0, 6),
        PackageStats("shared", 11, 9),
    ]

    print(f"{'package':<16} {'Ca':>4} {'Ce':>4} {'I':>8}  lecture")
    for package in packages:
        instability = package.instability()
        reading = "stable" if instability < 0.3 else "instable" if instability > 0.7 else "intermédiaire"
        print(f"{package.name:<16} {package.ca:>4} {package.ce:>4} {instability:>8.2f}  {reading}")
